lib/ecto/soft_delete_repo.ex

```diff
@@ -71,6 +111,61 @@ defmodule Ecto.SoftDelete.Repo do
         |> update!()
       end
 
+      def restore_all(struct, repo \\ :context) do
+        source = Ecto.get_meta(struct, :source)
+
+        context =
+          case repo do
+            :context -> Ecto.get_meta(struct, :context)
+            _ -> repo
+          end
+
+        %{
+          columns: _,
+          command: _,
+          connection_id: _,
+          messages: _,
+          num_rows: num_rows,
+          rows: rows
+        } =
+          Ecto.Adapters.SQL.query!(
+            context,
+            "UPDATE #{source} SET deleted_at = NULL"
+          )
+
+        {num_rows, rows}
+      end
+
+      def restore(struct, repo \\ :context, key \\ "id") do
+        value =
+          Map.from_struct(struct)
+          |> Map.get(String.to_atom(key))
+
+        source = Ecto.get_meta(struct, :source)
+
+        context =
+          case repo do
+            :context -> Ecto.get_meta(struct, :context)
+            _ -> repo
+          end
+
+        %{
+          columns: _,
+          command: _,
+          connection_id: _,
+          messages: _,
+          num_rows: num_rows,
+          rows: rows
+        } =
+          Ecto.Adapters.SQL.query!(
+            context,
+            "UPDATE #{source} SET deleted_at = NULL WHERE #{key} = $1",
+            [value]
+          )
+
+        {num_rows, rows}
+      end
```

Comment:
```suggestion
      def restore(struct_or_changeset) do
        struct_or_changeset
        |> Ecto.Changeset.change(deleted_at: nil)
        |> update()
      end
```

Pretty sure this will work

Reply:
SOOO after testing on your branch I was like : "WTF his test didn't work, WHY????"

And I found the reason

On your test you had

```elixir
    test "should soft restore the queryable" do
      user = Repo.insert!(%User{email: "test0@example.com"})
      # `user` = the inserted one
      Repo.soft_delete(user)
      # you delete the user BUT the var `user` still the inserted one

      # Now you try to remove the var `user` (remember it's the inserted one)
      Repo.restore(user)

      # ....
    end
```

So it does

```elixir
iex> Ecto.Changeset.change(%User{deleted_at: nil}, deleted_at: nil) # so no changes here
%Ecto.Changeset{changes: %{}}
```
> So Repo.update does nothing

Instead if you had assign `user = Repo.soft_delete!(user)` then you have

```elixir
iex> Ecto.Changeset.change(%User{deleted_at: ~U[2023-01-01 10:00:00]}, deleted_at: nil)
%Ecto.Changeset{changes: %{deleted_at: nil}}
```
> So Repo.update execute an update query



Sorry for long post I don't know you level or programming I prefer to be more exhaustive here

Reply:
ugh cant believe i missed that.